src/WSRequestHandler_General.cpp

```diff
@@ -281,6 +287,102 @@ RpcResponse WSRequestHandler::BroadcastCustomMessage(const RpcRequest& request)
 	return request.success();
 }
 
+/**
+ * Save user defined persistent data in OBS
+ *
+ * @param {String} `key` Identifier to be choosen by the client
+ * @param {String} `value` User-defined data
+ *
+ * @api general
+ * @name SetPersistentData
+ * @category general
+ * @since 4.7.0
+ */
+RpcResponse WSRequestHandler::SetPersistentData(const RpcRequest& request) {
+	if (!GetConfig()->PersistentDataEnabled) {
+		return request.failed("persistent data storage is disabled");
+	}
+
+	if (!request.hasField("key") || !request.hasField("value")) {
+		return request.failed("missing request parameters");
+	}
+
+	QString key = obs_data_get_string(request.parameters(), "key");
+	QString value = obs_data_get_string(request.parameters(), "value");
+
+	if (key.isEmpty()) {
+		return request.failed("key not specified!");
+	}
+
+	if (value.isEmpty()) {
+		return request.failed("value not specified!");
+	}
+
+	if (value.length() > GetConfig()->PersistentDataMaxSize) {
+		return request.failed("maximum data size exceeded, check configuration");
+	}
+
+	QString jsonString = GetConfig()->PersistentDataStore;
+	QJsonDocument jsonDoc = QJsonDocument::fromJson(jsonString.toUtf8());
```

Comment:
Requests can be processed on separate threads (even Get/SetPersistentData). It'd be best to throw a mutex in there to protect the data store from conflicts between two requests (e.g. two changes in parallel with the last discarding the first)

Reply:
You want to keep it in-memory as a QJsonbject. According to the documentation all functions of QJsonobject are reentrant.

https://doc.qt.io/qt-5/qjsondocument.html

> can also be called simultaneously from multiple threads, but only if each invocation uses its own data

So it should only fail if two clients write to the same key then and this I would perceive as more of a use error as even with a mutex this would result in unpredictable / discarded changes.

Reply:
From what I understand here, QJsonDocument's methods are reetrant because they don't rely on global/static data to function, apart from the JSON data owned by a QJsonDocument instance. However, the data structure used for JSON data in QJsonDocument is not guaranteed to be thread-safe, hence the need to protect it with a mutex.